astacus/coordinator/plugins/flink/steps.py

```diff
@@ -0,0 +1,134 @@
+"""
+Copyright (c) 2021 Aiven Ltd
+See LICENSE for details
+"""
+
+from astacus.common import ipc
+from astacus.common.asyncstorage import AsyncJsonStorage
+from astacus.common.exceptions import TransientException
+from astacus.common.utils import AstacusModel
+from astacus.coordinator.cluster import Cluster
+from astacus.coordinator.plugins.base import BackupManifestStep, Step, StepsContext
+from astacus.coordinator.plugins.flink.manifest import FlinkManifest
+from astacus.coordinator.plugins.zookeeper import ChangeWatch, ZooKeeperClient, ZooKeeperConnection
+from typing import Dict, List, Optional, Type
+
+import dataclasses
+import logging
+
+logger = logging.getLogger(__name__)
+
+
+@dataclasses.dataclass
+class RetrieveDataStep(Step):
+    """
+    Backups Flink tables from ZooKeeper.
+    """
+    zookeeper_client: ZooKeeperClient
+    zookeeper_paths: List[str]
+
+    async def run_step(self, cluster: Cluster, context: StepsContext):
+        async with self.zookeeper_client.connect() as connection:
+            change_watch = ChangeWatch()
+            results = {}
+            for zk_path in self.zookeeper_paths:
+                result = await self._get_result(zk_path, change_watch, connection)
+                if result:
+                    results.update(result)
+            if change_watch.has_changed:
+                raise TransientException("Concurrent modification during access entities retrieval")
+        return results
+
+    async def _get_result(self, zk_path, change_watch, connection):
+        result = {}
+        if await connection.exists(zk_path):
+            result[zk_path] = await self._get_data_recursively(
+                zk_prefix=zk_path, current_node=zk_path, connection=connection, watch=change_watch
+            )
+        return result
+
+    async def _get_data_recursively(
+        self, zk_prefix: str, current_node: str, connection: ZooKeeperConnection, watch: ChangeWatch
+    ):
+        result = {}
+        if await connection.exists(zk_prefix):
+            children = await connection.get_children(current_node, watch=watch)
+            for child in children:
+                value = await connection.get(f"{current_node}/{child}", watch=watch)
+                if value:
+                    result[child] = value.decode("UTF-8")
```

Comment:
Should we expect nodes with values that are not valid UTF-8 ?

(For ClickHouse that's a thing, but I don't know if Flink is entirely using UTF-8 text).

Reply:
Sorry, for answering for Anatoly - we store with Astacus only Aiven metadata, so we fully control what we are storing there. Flink related things, Flink decode and manage itself.